### ISSUE TITLE
rich.progress.track to work in Juypter Notebooks

### ISSUE THREAD

**Issue by leej11** [willmcgugan/rich#131]
Hey thanks so much for this package! Just discovered it and have been playing around with it this morning. 

I quite often work in JupyterLab / Jupyter Notebooks and most of the functionality seems to work and print fine as expected.

However, when trying to employ the `rich.progress.track` functionality, or alternatively the `rich.progress.Progress` class, I can't get it to print the progress bar as output in the Jupyter Notebook.

So here's the code I've been trying:

```python
from rich.progress import track
import time

total = 0

def a_medium_speed_func(number):
    time.sleep(0.5)
    global total
    total += number

for i in track(range(100), description="Processing..."):
    a_medium_speed_func(i)
```

It's a super dumb function but just made it to see the progress bar working.

I dug around and noticed a similar issue had been raised #26 and so I tried searching the docs and did find something that looked promising... [rich.console.Console - force_jupyer](https://rich.readthedocs.io/en/latest/reference/console.html#rich.console.Console) so I tried the following code:

```python
from rich.progress import track
from rich.console import Console
import time

console = Console(force_jupyter=True)

total = 0

def a_medium_speed_func(number):
    time.sleep(0.5)
    global total
    total += number

for i in track(range(100), description="Processing...", console=console):
    a_medium_speed_func(i)
```

But that still didn't work. I'm not sure if it is something I am coding wrong, or perhaps there is missing functionality so I thought I'd raise this issue. 

If there's any more details you'd need from me let me know! I'm happy to help in whatever way I can.

**Comment by willmcgugan** [comment]
Progress bars don't work with jupyter I'm afraid (`force_jupyter` overrides jupyter auto-detect).

There's more work required to make a working progress bar in Jupyter. It would be a nice to have, but its not high on my list of priorities tbh. I'll tackle it eventually. Maybe sooner if I get sponsorship!

**Comment by staubda** [comment]
Just wanted to add a +1 for this feature request, fingers crossed for your funding!

**Comment by willmcgugan** [comment]
I've got Jupyter progress bars working. It's still at the experimental stage. Early feedback would be great, if you could check out [this branch](https://github.com/willmcgugan/rich/tree/jupyter-progress) and let me know how that works for you.

The interaction with other output is different with Jupyter. Rather than display the progress beneath the output, it is displayed above it. This seems to be the only practical solution with Jupyter.

**Comment by willmcgugan** [comment]
Please give v3.1.0 a try which has progress support for Jupyter notebooks.

**Comment by leej11** [comment]
Hi @willmcgugan just tried this out today.

![image](https://user-images.githubusercontent.com/17391902/87163386-2bf15600-c2bf-11ea-92bd-26977d2dfb62.png)

First time I ran I got `ModuleNotFoundError: No module named 'ipywidgets'`, but after pip installing it worked perfectly. Does that mean ipywidgets needs to be added as a dependency I guess? I would attempt to offer a PR for that but I'm used to seeing either `environment.yml` or `requirements.txt` but it looks like you're using `poetry` ?

![image](https://user-images.githubusercontent.com/17391902/87163400-34e22780-c2bf-11ea-9141-a41d6703fefc.png)

Anyway, looks very nice!! Can't wait to integrate it into some stuff.
